test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/FhirStorageTests.cs

```diff
@@ -67,6 +68,78 @@ public FhirStorageTests(FhirStorageTestsFixture fixture)
+        [FhirStorageTestsFixtureArgumentSets(DataStore.SqlServer)]
+        public async Task TimeTravel()
+        {
+            await _fixture.SqlHelper.ExecuteSqlCmd("DELETE FROM dbo.Resource"); // remove all data
+
+            // add resource
+            var type = "Patient";
+            var patient = (Patient)Samples.GetJsonSample(type).ToPoco();
+            patient.Id = Guid.NewGuid().ToString();
+            await Mediator.UpsertResourceAsync(patient.ToResourceElement());
+
+            await Task.Delay(100); // avoid time -> surrogate id -> time round trip error
+            var till = DateTime.UtcNow;
+
+            var results = await _fixture.SearchService.SearchAsync(type, new List<Tuple<string, string>>(), CancellationToken.None);
+            Assert.Single(results.Results);
+            var resource = results.Results.First().Resource;
+            Assert.Equal("1", resource.Version);
+
+            // add till and check that resource is returned
+            var queryParameters = new[] { Tuple.Create(KnownQueryParameterNames.LastUpdated, $"le{new PartialDateTime(till)}") };
+            results = await _fixture.SearchService.SearchAsync(type, queryParameters, CancellationToken.None);
+            Assert.Single(results.Results);
+            resource = results.Results.First().Resource;
+            Assert.Equal("1", resource.Version);
+            Assert.False(resource.IsHistory); // current
+
+            await UpdateResource(patient); // update resource
+
+            // !!! pre time travel behavior
+            // same parameters
+            // resource is not returned because it became "invisible" in the time interval requested
+            results = await _fixture.SearchService.SearchAsync(type, queryParameters, CancellationToken.None);
+            Assert.Empty(results.Results);
+
+            // add magic parameters
+            var range = (await _fixture.SearchService.GetSurrogateIdRanges(type, DateTime.MinValue, till, 1, CancellationToken.None)).First();
+            queryParameters = new[]
+            {
+                Tuple.Create(KnownQueryParameterNames.Type, type),
+                Tuple.Create(KnownQueryParameterNames.GlobalEndSurrogateId, range.GlobalEnd.ToString()),
+                Tuple.Create(KnownQueryParameterNames.EndSurrogateId, range.End.ToString()),
+                Tuple.Create(KnownQueryParameterNames.GlobalStartSurrogateId, range.GlobalStart.ToString()),
+                Tuple.Create(KnownQueryParameterNames.StartSurrogateId, range.Start.ToString()),
+            };
+
+            // !!! time travel behavior
+            results = await _fixture.SearchService.SearchAsync(type, queryParameters, CancellationToken.None);
+            Assert.Single(results.Results);
+            resource = results.Results.First().Resource;
+            Assert.Equal("1", resource.Version);
+            Assert.False(resource.IsHistory); // it is returned as current but is marked as history in the database ???
```

Comment:
This is strange. We should look into why this is and think about if this is the right behavior.